pravega/pravega
# Issue 2513: Client side versioned binary serialization

**Change log description**
Add support for versioned serialization (#1956) for the public facing java client objects.

**Purpose of the change**
Currently rely on Java serialization for classes that are passed to the application and which need to be passed back. This forces users to depend on Java serialization which is less than ideal. 
Additionally we need to support versioning of these objects, as they will by definition persist through client upgrades. 

**What the code does**
To achieve this a new pattern is introduced. On each of these classes there is a `toBytes()` method that will serialize the object. Additionally there is a `fromBytes()` method on the public interface that takes the serialized bytes and re-instantiates the the original object. 

This fixes #2513, by using the same serialization framework developed to support the Server in #1956. This provides a way to modify the data going forward while preserving backwards compatibility. 

This change does not yet address: Config files, or the ReaderGroupState. We may wish to change those over at a future time, but for now they remain Java serializable. 

It should be noted that **This is a breaking change, and that client libraries with this code will not be compatible with serialized objects from previous versions of the client.**

This also incidentally fixes #2502, #2490, #256 and #1941

**How to verify it**
There are tests added to verify that each of the new serialization paths, round trips the data. 
Existing integration tests confirm that the objects that are no longer serializable don't affect the readerGroupState and the config objects.
